Agent container crash after unsuccessful model spec unmarshal

/kind bug

**What steps did you take and what happened:**

It happened to me, that agent syncer failed to unmarshal model spec in kserve/pkg/agent/syncer.go:62
When this happens SyncModelDir return nil and later agent crashes with the following logs&stacktrace:

```
{"severity":"INFO","timestamp":"2022-05-31T11:08:48.624769564Z","caller":"agent/syncer.go:43","message":"Syncing from model success file SUCCESS.4ce6dba39616657ab44effb6d207b4c2eb3696dfb20e2241b0be201220eb8c46"}
{"severity":"ERROR","timestamp":"2022-05-31T11:08:48.624834324Z","caller":"agent/watcher.go:43","message":"Failed to sync model dir error in syncing model dir: failed to unmarshal model spec: unexpected end of JSON input","stacktrace":"github.com/kserve/kserve/pkg/agent.NewWatcher\n\t/go/src/github.com/kserve/kserve/pkg/agent/watcher.go:43\nmain.startModelPuller\n\t/go/src/github.com/kserve/kserve/cmd/agent/main.go:266\nmain.main\n\t/go/src/github.com/kserve/kserve/cmd/agent/main.go:117\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:255"}
panic: assignment to entry in nil map

goroutine 1 [running]:
github.com/kserve/kserve/pkg/agent.(*Watcher).parseConfig(0xc0009919b8, {0xc000292000, 0x4c0, 0x1769fe0}, 0xb0)
	/go/src/github.com/kserve/kserve/pkg/agent/watcher.go:139 +0x677
github.com/kserve/kserve/pkg/agent.(*Watcher).syncModelConfig(0x1a94eb9, {0xc0000c7d28, 0xc0002d1970}, 0x2)
	/go/src/github.com/kserve/kserve/pkg/agent/watcher.go:75 +0xd5
github.com/kserve/kserve/pkg/agent.NewWatcher({0x7fffa54a175a, 0xc}, {0x7fffa54a1773, 0x0}, 0xc0006406d8)
	/go/src/github.com/kserve/kserve/pkg/agent/watcher.go:53 +0x23f
main.startModelPuller(0xc0006406d8)
	/go/src/github.com/kserve/kserve/cmd/agent/main.go:266 +0x11c
main.main()
	/go/src/github.com/kserve/kserve/cmd/agent/main.go:117 +0x25f
```

Failure in SyncModelDir leads to panic later in the code, because modelTracker is nil. This leads to crash loop in the agent container.


**What did you expect to happen:**

When agent fails to unmarshal model it should just report the error and not crash.

**Anything else you would like to add:**

I suggest fixing this by not returning nil from SyncModelDir. Instead I suggest returning empty modelTracker. Here is the diff of the fix:

```
diff --git a/pkg/agent/syncer.go b/pkg/agent/syncer.go
index 3471462a..5fb3918a 100644
--- a/pkg/agent/syncer.go
+++ b/pkg/agent/syncer.go
@@ -71,7 +72,8 @@ func SyncModelDir(modelDir string, logger *zap.SugaredLogger) (map[string]modelW
                return nil
        })
        if err != nil {
-               return nil, errors.Wrapf(err, "error in syncing model dir")
+               return modelTracker, errors.Wrapf(err, "error in syncing model dir")
        }
        return modelTracker, nil
 }
```

**Environment:**


- KFServing Version: latest (built from master)
- Kubernetes version: (use `kubectl version`): 1.21

--- a/pkg/agent/syncer.go
+++ b/pkg/agent/syncer.go
@@ -71,7 +71,7 @@ func SyncModelDir(modelDir string, logger *zap.SugaredLogger) (map[string]modelW
 		return nil
 	})
 	if err != nil {
-		return nil, errors.Wrapf(err, "error in syncing model dir")
+		return modelTracker, errors.Wrapf(err, "error in syncing model dir")
 	}
 	return modelTracker, nil
 }

--- a/pkg/agent/watcher_test.go
+++ b/pkg/agent/watcher_test.go
@@ -67,6 +67,20 @@ var _ = Describe("Watcher", func() {
 	})
 
 	Describe("Sync models config on startup", func() {
+		It("should not crash after sync error", func() {
+			badModelDir := filepath.Join(modelDir, "bad-model")
+			configDir := filepath.Join(modelDir, "configs")
+			Expect(os.MkdirAll(badModelDir, os.ModePerm)).To(Succeed())                                       //nolint:gosec // G301: test directory
+			Expect(os.MkdirAll(configDir, os.ModePerm)).To(Succeed())                                         //nolint:gosec // G301: test directory
+			Expect(os.WriteFile(filepath.Join(badModelDir, "SUCCESS.bad"), []byte("{"), 0o644)).To(Succeed()) //nolint:gosec // G306: test file
+			modelConfigs := modelconfig.ModelConfigs{{Name: "model1"}}
+			data, err := json.Marshal(modelConfigs)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(os.WriteFile(filepath.Join(configDir, constants.ModelConfigFileName), data, 0o644)).To(Succeed()) //nolint:gosec // G306: test file
+			Expect(func() {
+				NewWatcher(configDir, modelDir, sugar)
+			}).ToNot(Panic())
+		})
 		Context("Getting new model events", func() {
 			It("should download and load the new models", func() {
 				logger.Printf("Sync model config using temp dir %v\n", modelDir)

